### PR TITLE
isShared cache

### PR DESCRIPTION
**Pull request by kevodwyer**

### CODE REVIEW

File: src/peergos/shared/user/UserContext.java [hunk 1]
```diff
@@ -30,6 +30,7 @@
     public final SigningKeyPair signer;
     public final BoxingKeyPair boxer;
     public final Fragmenter fragmenter;
+    public final Set<String> sharedFiles;
```
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: Would be good to know what these strings are. Ideally with a type. Maybe even a class to wrap the whole set too.

File: src/peergos/shared/user/UserContext.java [hunk 2]
```diff
@@ -648,6 +621,59 @@ public boolean isJavascript() {
...
+                    return shareWithAll(sharedPath, remainingReaders);
+                });
+
+    }
+
+    private String shareKey(FileTreeNode file) {
+        if(file == null) {
+            return "";
+        }
+        String key = file.getPointer().filePointer.toString();
```
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: Should use the location.

File: src/peergos/shared/user/UserContext.java [hunk 3]
```diff
@@ -31,6 +31,7 @@
     public final SigningPrivateKeyAndPublicHash signer;
     public final BoxingKeyPair boxer;
     public final Fragmenter fragmenter;
+    public final Set<String> sharedFiles;
```
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: maybe call this filesSharedWithFriends

File: src/peergos/shared/user/fs/FileTreeNode.java [hunk 4]
```diff
@@ -233,7 +233,7 @@ public String toLink() {
 
     @JsMethod
     public boolean isWritable() {
-        return entryWriterKey.isPresent();
+        return entryWriterKey == null ? false : entryWriterKey.isPresent();
```
Comment (ianopolous) [inline src/peergos/shared/user/fs/FileTreeNode.java]: add a comment as to why this is needed

File: src/peergos/shared/user/UserContext.java [hunk 5]
```diff
@@ -926,8 +996,49 @@ private FollowRequest decodeFollowRequest(byte[] raw) throws IOException {
...
+                    return Futures.combineAll(childrenFTNs.stream()
+                            .map(childFTN -> sharedWith(childFTN, sharedUserDirs))
+                            .collect(Collectors.toList()))
+                            .thenApply(x -> childrenFTNs);
+                });
+            });
+        });
+    }
+
+    private CompletableFuture<Boolean> sharedWith(FileTreeNode file, Set<FileTreeNode> sharedUserDirs) {
```
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: rename to something more meaningful

File: src/peergos/shared/user/UserContext.java (hunk 3 above)
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: and some indication as to the type of the String (that it is a Location)

Comment (kevodwyer) [inline src/peergos/shared/user/UserContext.java]: done

File: src/peergos/shared/user/fs/FileTreeNode.java (hunk 4 above)
Comment (kevodwyer) [inline src/peergos/shared/user/fs/FileTreeNode.java]: done

File: src/peergos/shared/user/UserContext.java (hunk 5 above)
Comment (kevodwyer) [inline src/peergos/shared/user/UserContext.java]: done

**Comment by kevodwyer** [comment]
ready for review yet again.

**Comment by ianopolous** [comment]
3 tests seems to be failing in MultiUserTests

**Pull request closed without merging**